tensorflow/c/kernels_experimental.cc

```diff
@@ -194,6 +195,33 @@ void TF_AssignVariable(TF_OpKernelContext* ctx, int input_index,
   TF_SetStatus(status, TF_OK, "");
 }
 
+void TF_AssignRefVariable(TF_OpKernelContext* ctx, int input_ref_index,
+                          int output_ref_index, int value_index,
+                          bool use_locking, bool validate_shape,
+                          void (*copyFunc)(TF_OpKernelContext* ctx,
+                                           TF_Tensor* source, TF_Tensor* dest),
+                          TF_Status* status) {
+  auto* cc_ctx = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+
+  auto copy = [copyFunc, ctx](::tensorflow::OpKernelContext* cc_ctx,
+                              ::tensorflow::Tensor* lhs,
+                              const ::tensorflow::Tensor& rhs) {
+    ::tensorflow::Status s;
+    TF_Tensor* tf_lhs = TF_TensorFromTensor(*lhs, &s);
```

Comment:
@wangpengmit What are the desired ownership semantics here? I saw that `TF_AssignVariable` transfers ownership to the caller, but it's not mentioned anywhere in the header so callers have no way to know whether they should delete the tensor. Do we want TensorFlow to keep ownership of the tensor (in which case we should call TF_DeleteTensor after `copyFunc`), or do we want to leave it as-is and let the pluggable device delete the tensor (in which case we should mention it in the header)?

Reply:
I agree with your analysis. It's a tough call. My biggest concern is consistency with other APIs in this file. I saw that they all (implicitly) rely on the callbacks to delete the `TF_Tensor`s, so maybe it's too late to change that convention (which definitely should be documented). OTOH those APIs are still in experimental status.

My suggestion is to align the new API with the existing APIs, and file a bug to require changing all APIs to not rely on callbacks to delete `TF_Tensor`s. We'll need to be careful when fixing the bug, communicating the change to existing users.

@penpornk I'm curious whether users of existing APIs already suffer from memory leaks because of this undocumented contract.

Reply:
We had memory leaks in some models because we didn't think we had to manage it ourselves, so there's at least one case ^^

Since the header is in experimental state, I don't think anyone should expect it to work indefinitely without breaking backward compatibility. But I also agree that all APIs should be aligned, so following the existing convention is probably safest for now.

Reply:
@wangpengmit TF-Metal plug-in and Intel Extension for TensorFlow (ITEX) plug-in could be using these kernel experimental APIs.
* TF-Metal (@kulinseth) proposed and implemented the API so I don't think they are experiencing leaks.
* ITEX: @Jianhui-Li, @jzhoulon, and @yiqianglee, have you been using the kernel experimental APIs? If so, it might be worth checking for leaks.

Reply:
@PatriceVignola OK, then please add a docstring in `kernels_experimental.h` saying that all callbacks in that file need to delete the `TF_Tensor`s. We'll try changing the contract later, which will be a big PR by itself. (It will be appreciated if you can file a bug for this.)

Reply:
I went ahead and added docstrings for all functions that take a callback. I also opened a bug: https://github.com/tensorflow/tensorflow/issues/55659

Reply:
Thanks!